<center> <img src = https://raw.githubusercontent.com/AndreyRysistov/DatasetsForPandas/main/hh%20label.jpg alt="drawing" style="width:400px;">

# <center> Проект: Анализ вакансий из HeadHunter
   

In [154]:
import pandas as pd
import psycopg2

In [143]:
DBNAME = 'project_sql'
USER = 'skillfactory'
PASSWORD = 'cCkxxLVrDE8EbvjueeMedPKt'
HOST = '84.201.134.129'
PORT = 5432

In [163]:
connection = psycopg2.connect(
    dbname=DBNAME,
    user=USER,
    host=HOST,
    password=PASSWORD,
    port=PORT
)

# Юнит 3. Предварительный анализ данных

1. Напишите запрос, который посчитает количество вакансий в нашей базе (вакансии находятся в таблице vacancies). 

In [4]:

query = f'''SELECT
COUNT(id) AS "Количество вакансий"

FROM public.vacancies 
'''


In [5]:
df = pd.read_sql_query(query, connection)
df




C:\Users\COMP\AppData\Local\Temp\ipykernel_24600\1455829647.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_1, connection)


Количество вакансий
0                49197

2. Напишите запрос, который посчитает количество работодателей (таблица employers). 

In [7]:
query = f'''SELECT
COUNT(id) AS "Количество работодателей"

FROM public.EMPLOYERS
'''

In [8]:
df = pd.read_sql_query(query, connection)
df

C:\Users\COMP\AppData\Local\Temp\ipykernel_24600\3851344990.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_1, connection)


Количество работодателей
0                     23501

3. Посчитате с помощью запроса количество регионов (таблица areas).

In [9]:
query = f'''SELECT
COUNT(id) AS "Количество регионов"

FROM public.AREAS
'''

In [10]:
df = pd.read_sql_query(query, connection)
df

C:\Users\COMP\AppData\Local\Temp\ipykernel_24600\3851344990.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_1, connection)


Количество регионов
0                 1362

4. Посчитате с помощью запроса количество сфер деятельности в базе (таблица industries).

In [11]:
query = f'''SELECT
COUNT(id) AS "Количество сфер деятельности"

FROM public.INDUSTRIES
'''

In [150]:
df=pd.read_sql_query(query, connection)
df

C:\Users\COMP\AppData\Local\Temp\ipykernel_24600\2741276783.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df=(pd.read_sql_query(query_3_1, connection))


Работодатель  Средняя зарплата             Название вакансии  \
0         Итон          450000.0  ML Engineer \ Data scientist   

  Тип рабочего графика Тип трудоустройства Требование к опыту  Регион  
0          Полный день    Полная занятость      От 3 до 6 лет  Москва

***

Выявил количество вакасий - 49197, выявил количество работодателей - 23501, посчитал количество регионов - 1362, посчитал количество сфер деятельности - 294

# Юнит 4. Детальный анализ вакансий

1. Напишите запрос, который позволит узнать, сколько (cnt) вакансий в каждом регионе (area).
Отсортируйте по количеству вакансий в порядке убывания.

In [30]:
query = f'''SELECT
a.name AS "Регион",
COUNT(v.id) AS "Количество вакансий"

FROM public.VACANCIES AS v
    JOIN public.AREAS AS a
    ON v.area_id = a.id
    
GROUP BY a.name
ORDER BY "Количество вакансий" DESC
'''

In [31]:
df = pd.read_sql_query(query, connection)
df

C:\Users\COMP\AppData\Local\Temp\ipykernel_24600\3851344990.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_1, connection)


Регион  Количество вакансий
0             Москва                 5333
1    Санкт-Петербург                 2851
2              Минск                 2112
3        Новосибирск                 2006
4             Алматы                 1892
..               ...                  ...
764       Тарко-Сале                    1
765    Новоаннинский                    1
766            Бирск                    1
767           Сасово                    1
768             Анна                    1

[769 rows x 2 columns]

2. Напишите запрос, чтобы определить у какого количества вакансий заполнено хотя бы одно из двух полей с зарплатой.

In [20]:
query = f'''SELECT
COUNT(id)

FROM public.VACANCIES

WHERE salary_from IS NOT NULL or salary_to IS NOT NULL
'''

In [21]:
df = pd.read_sql_query(query, connection)
df

C:\Users\COMP\AppData\Local\Temp\ipykernel_1364\3851344990.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_1, connection)


count
0  24073

3. Найдите средние значения для нижней и верхней границы зарплатной вилки. Округлите значения до целого.

In [22]:
query = f'''SELECT
ROUND(AVG(salary_from)) AS "Средняя з/п нижней границы вилки",
ROUND(AVG(salary_to)) AS "Средняя з/п верхней границы вилки"

FROM public.VACANCIES
'''

In [23]:
df = pd.read_sql_query(query, connection)
df

C:\Users\COMP\AppData\Local\Temp\ipykernel_1364\3851344990.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_1, connection)


Средняя з/п нижней границы вилки  Средняя з/п верхней границы вилки
0                           71065.0                           110537.0

4. Напишите запрос, который выведет количество вакансий для каждого сочетания типа рабочего графика (schedule) и типа трудоустройства (employment), используемого в вакансиях. Результат отсортируйте по убыванию количества.


In [24]:
query = f'''SELECT
schedule||' — '||employment AS "Сочетание",
COUNT(schedule||' — '||employment) AS "Количество"

FROM public.VACANCIES

GROUP BY employment, schedule
ORDER BY "Количество" DESC
'''

In [25]:
df = pd.read_sql_query(query, connection)
df

C:\Users\COMP\AppData\Local\Temp\ipykernel_1364\3851344990.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_1, connection)


Сочетание  Количество
0           Полный день — Полная занятость       35367
1      Удаленная работа — Полная занятость        7802
2         Гибкий график — Полная занятость        1593
3   Удаленная работа — Частичная занятость        1312
4        Сменный график — Полная занятость         940
5                 Полный день — Стажировка         569
6        Вахтовый метод — Полная занятость         367
7        Полный день — Частичная занятость         347
8      Гибкий график — Частичная занятость         312
9           Полный день — Проектная работа         141
10     Удаленная работа — Проектная работа         133
11              Гибкий график — Стажировка         116
12    Сменный график — Частичная занятость         101
13           Удаленная работа — Стажировка          64
14        Гибкий график — Проектная работа          18
15             Сменный график — Стажировка          12
16       Вахтовый метод — Проектная работа           2
17       Сменный график — Проектная работа           1

5. Напишите запрос, выводящий значения поля Требуемый опыт работы (experience) в порядке возрастания количества вакансий, в которых указан данный вариант опыта. 

In [27]:
query = f'''SELECT
experience AS "Требования к опыту",
COUNT(id) "Количество вакансий"

FROM public.VACANCIES

GROUP BY experience
ORDER BY "Количество вакансий" 
'''

In [28]:
df = pd.read_sql_query(query, connection)
df

C:\Users\COMP\AppData\Local\Temp\ipykernel_1364\3851344990.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_1, connection)


Требования к опыту  Количество вакансий
0         Более 6 лет                 1337
1           Нет опыта                 7197
2       От 3 до 6 лет                14511
3  От 1 года до 3 лет                26152

***

В Москве больше всего вакансий.Не всегда указана зарплата. В 24073 акансий казана зарплата.Средняя з/п нижней границы вилки равна 71065 рублей, а Средняя з/п верхней границы вилки равна 110537 рублей.Самый популярный режим работы это Полный день — Полная занятость.Меньше всего вакансий с опытом работы более 6 лет, больше всего с опытом от 1 года до 3 лет.

# Юнит 5. Анализ работодателей

1. Напишите запрос, который позволит узнать, какие работодатели находятся на первом и пятом месте по количеству вакансий.

In [35]:
query = f'''(SELECT
e.name AS "Наименование работодателя",
COUNT (v.id) AS "Количество вакансий"

    FROM public.VACANCIES AS v
    JOIN public.EMPLOYERS AS e
    ON e.id = v.employer_id
    
GROUP BY e.name
ORDER BY 2 desc
LIMIT 1)

UNION ALL

(SELECT
e.name,
COUNT(v.id)

    FROM public.VACANCIES AS v
    JOIN public.EMPLOYERS AS e
    ON e.id = v.employer_id
    
GROUP BY e.name
ORDER BY 2 DESC
OFFSET 4 LIMIT 1)
'''

In [38]:
df = pd.read_sql_query(query, connection)
df

C:\Users\COMP\AppData\Local\Temp\ipykernel_24600\3851344990.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_1, connection)


Наименование работодателя  Количество вакансий
0                    Яндекс                 1933
1             Газпром нефть                  331

2. Напишите запрос, который для каждого региона выведет количество работодателей и вакансий в нём.
Среди регионов, в которых нет вакансий, найдите тот, в котором наибольшее количество работодателей.


In [38]:
query = f'''SELECT
a.name AS "Регион",
COUNT(v.id) AS "Количество вакансий",
COUNT(e.id) AS "Количество работодателей"

FROM public.AREAS a
    LEFT JOIN public.EMPLOYERS e
    ON a.id = e.area

    LEFT JOIN public.vacancies v
    ON a.id = v.area_id

WHERE v.id IS NULL
GROUP BY a.name
ORDER BY "Количество работодателей" DESC
LIMIT 1
'''

In [39]:
df = pd.read_sql_query(query, connection)
df

C:\Users\COMP\AppData\Local\Temp\ipykernel_1364\3851344990.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_1, connection)


Регион  Количество вакансий  Количество работодателей
0  Россия                    0                       410

3. Для каждого работодателя посчитайте количество регионов, в которых он публикует свои вакансии. Отсортируйте результат по убыванию количества.


In [40]:
query = f'''SELECT
e.name AS "Работодатель",
COUNT(distinct v.area_id) AS "Количество регионов"

FROM public.EMPLOYERS e  
    JOIN public.VACANCIES v
    ON v.employer_id = e.id
    
GROUP BY e.name
ORDER BY "Количество регионов" DESC

'''

In [41]:
df = pd.read_sql_query(query, connection)
df

C:\Users\COMP\AppData\Local\Temp\ipykernel_1364\3851344990.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_1, connection)


Работодатель  Количество регионов
0                      Яндекс                  181
1                  Ростелеком                  152
2                  Спецремонт                  116
3      Поляков Денис Иванович                   88
4                   ООО ЕФИН                    71
...                       ...                  ...
14761                  UniSol                    1
14762            UNISTORY LLC                    1
14763                   UNIT6                    1
14764     United Distribution                    1
14765              UNITEL LLC                    1

[14766 rows x 2 columns]

4. Напишите запрос для подсчёта количества работодателей, у которых не указана сфера деятельности. 

In [44]:
query = f'''SELECT
COUNT(e.id) AS "Количество работодателей"

FROM public.EMPLOYERS_INDUSTRIES AS EI  
    RIGHT OUTER JOIN public.EMPLOYERS e
    ON EI.employer_id = e.id
    
    LEFT JOIN public.INDUSTRIES i
    ON EI.industry_id = i.id 
WHERE i.name IS NULL

'''

In [45]:
df = pd.read_sql_query(query, connection)
df

C:\Users\COMP\AppData\Local\Temp\ipykernel_1364\3851344990.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_1, connection)


Количество работодателей
0                      8419

5. Напишите запрос, чтобы узнать название компании, находящейся на третьем месте в алфавитном списке (по названию) компаний, у которых указано четыре сферы деятельности. 

In [46]:
query = f'''SELECT
e.name AS "Название компании"

FROM public.EMPLOYERS_INDUSTRIES AS EI  
    RIGHT OUTER JOIN public.EMPLOYERS e
    ON EI.employer_id = e.id

    LEFT JOIN public.INDUSTRIES i
    ON EI.industry_id = i.id
    
GROUP BY e.name
HAVING COUNT(i.name) = 4
ORDER BY e.name
OFFSET 2 LIMIT 1

'''


In [47]:
df = pd.read_sql_query(query, connection)
df

C:\Users\COMP\AppData\Local\Temp\ipykernel_1364\3851344990.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_1, connection)


Название компании
0              2ГИС

6. С помощью запроса выясните, у какого количества работодателей в качестве сферы деятельности указана Разработка программного обеспечения.


In [48]:
query = f'''SELECT
COUNT(e.name) AS "Количество работодателей"

FROM public.EMPLOYERS_INDUSTRIES AS EI  
    RIGHT OUTER JOIN public.EMPLOYERS e
    ON EI.employer_id = e.id
    
    LEFT JOIN public.INDUSTRIES i
    ON EI.industry_id = i.id 
    
WHERE i.name = 'Разработка программного обеспечения'

'''

In [49]:
df = pd.read_sql_query(query, connection)
df

C:\Users\COMP\AppData\Local\Temp\ipykernel_1364\3851344990.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_1, connection)


Количество работодателей
0                      3553

7. Для компании «Яндекс» выведите список регионов-миллионников, в которых представлены вакансии компании, вместе с количеством вакансий в этих регионах. Также добавьте строку Total с общим количеством вакансий компании. Результат отсортируйте по возрастанию количества.

Список городов-милионников надо взять [отсюда](https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8). 

Если возникнут трудности с этим задание посмотрите материалы модуля  PYTHON-17. Как получать данные из веб-источников и API. 

In [33]:
cities = pd.read_html('https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8')
city = tuple(cities[1]['Город'])
city


('Москва',
 'Санкт-Петербург',
 'Новосибирск',
 'Екатеринбург',
 'Казань',
 'Нижний Новгород',
 'Красноярск',
 'Челябинск',
 'Самара',
 'Уфа',
 'Ростов-на-Дону',
 'Краснодар',
 'Омск',
 'Воронеж',
 'Пермь',
 'Волгоград')

In [34]:
query = f'''with VACANCIES_AREAS as (SELECT
a.name AS "Название региона",
COUNT(v.id) AS "Количество вакансий"

FROM public.VACANCIES AS v
    JOIN public.EMPLOYERS AS e 
    ON e.id =  v.employer_id
    
    JOIN public.AREAS AS a 
    ON a.id =  v.area_id
    
WHERE e.name = 'Яндекс' AND a.name in {city}
GROUP BY a.name)

(SELECT
va."Название региона",
va."Количество вакансий"
FROM VACANCIES_AREAS AS va)

UNION ALL

(SELECT
'total',
SUM("Количество вакансий") AS "Сумма вакансий"
FROM VACANCIES_AREAS)
ORDER BY "Количество вакансий"

'''

In [65]:
df = pd.read_sql_query(query, connection)
df

C:\Users\COMP\AppData\Local\Temp\ipykernel_1364\3851344990.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_1, connection)


Название региона  Количество вакансий
0              Омск                 21.0
1         Челябинск                 22.0
2        Красноярск                 23.0
3         Волгоград                 24.0
4             Пермь                 25.0
5            Казань                 25.0
6    Ростов-на-Дону                 25.0
7            Самара                 26.0
8               Уфа                 26.0
9         Краснодар                 30.0
10          Воронеж                 32.0
11      Новосибирск                 35.0
12  Нижний Новгород                 36.0
13     Екатеринбург                 39.0
14  Санкт-Петербург                 42.0
15           Москва                 54.0
16            total                485.0

***

На первом месте по числу вакансий находится Яндекс. В Росиии 410 работадатеей не разместили вакансий.Яндекс разместил вакансии в 181 регионе. Работодатели в количестве 8419 не указали сферу дятельности. Больше всего вакансий от работодателя Яндекс в Москве.

# Юнит 6. Предметный анализ

1. Сколько вакансий имеет отношение к данным?

Считаем, что вакансия имеет отношение к данным, если в её названии содержатся слова 'data' или 'данн'.

*Подсказка: Обратите внимание, что названия вакансий могут быть написаны в любом регистре.* 


In [4]:
query = f'''SELECT
COUNT(name) AS "Количество вакансий"

FROM public.VACANCIES 
WHERE (lower(name) LIKE '%data%') OR (lower(name) LIKE '%данн%')

'''

In [5]:
df = pd.read_sql_query(query, connection)
df

C:\Users\COMP\AppData\Local\Temp\ipykernel_19600\3851344990.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_1, connection)


Количество вакансий
0                 1771

2. Сколько есть подходящих вакансий для начинающего дата-сайентиста? 
Будем считать вакансиями для дата-сайентистов такие, в названии которых есть хотя бы одно из следующих сочетаний:
* 'data scientist'
* 'data science'
* 'исследователь данных'
* 'ML' (здесь не нужно брать вакансии по HTML)
* 'machine learning'
* 'машинн%обучен%'

** В следующих заданиях мы продолжим работать с вакансиями по этому условию.*

Считаем вакансиями для специалистов уровня Junior следующие:
* в названии есть слово 'junior' *или*
* требуемый опыт — Нет опыта *или*
* тип трудоустройства — Стажировка.
 

In [6]:
query = f'''SELECT
COUNT(name) AS "Количество вакансий"

FROM public.VACANCIES 

WHERE (lower(name) LIKE '%data scientist%' 
OR lower(name) LIKE'%data science%'
OR lower(name) LIKE '%исследователь данных%'
OR (name LIKE '%ML%' and name NOT LIKE '%HTML%')
OR lower(name) LIKE '%machine learning%'
OR lower(name) LIKE '%машинн%обучен%') 
AND (lower(name) LIKE '%junior%'
OR experience LIKE '%Нет опыта%'
OR employment LIKE '%Стажировка%')

'''



In [7]:
df = pd.read_sql_query(query, connection)
df 

C:\Users\COMP\AppData\Local\Temp\ipykernel_19600\559104121.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_1, connection)


Количество вакансий
0                   51

3. Сколько есть вакансий для DS, в которых в качестве ключевого навыка указан SQL или postgres?

** Критерии для отнесения вакансии к DS указаны в предыдущем задании.*

In [8]:
query = f'''SELECT
COUNT(name) AS "Количество вакансий"

FROM public.VACANCIES 

WHERE (lower(name) LIKE '%data scientist%' 
OR lower(name) LIKE '%data science%'
OR lower(name) LIKE '%исследователь данных%'
OR (name LIKE '%ML%' AND name NOT LIKE '%HTML%')
OR lower(name) LIKE '%machine learning%'
OR lower(name) LIKE '%машинн%обучен%')
AND (key_skills LIKE '%SQL%' OR key_skills LIKE '%postgres%')

'''

In [9]:
df = pd.read_sql_query(query, connection)
df 

C:\Users\COMP\AppData\Local\Temp\ipykernel_19600\559104121.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_1, connection)


Количество вакансий
0                  201

4. Проверьте, насколько популярен Python в требованиях работодателей к DS.Для этого вычислите количество вакансий, в которых в качестве ключевого навыка указан Python.

** Это можно сделать помощью запроса, аналогичного предыдущему.*

In [10]:
query = f'''SELECT
COUNT(name) "Количество вакансий"

FROM public.VACANCIES 

WHERE (lower(name) LIKE '%data scientist%' 
OR lower(name) LIKE '%data science%'
OR lower(name) LIKE '%исследователь данных%'
OR (name LIKE '%ML%' and name NOT LIKE '%HTML%')
OR lower(name) LIKE '%machine learning%'
OR lower(name) LIKE '%машинн%обучен%')
AND key_skills LIKE '%Python%'

'''

In [11]:
df = pd.read_sql_query(query, connection)
df 

C:\Users\COMP\AppData\Local\Temp\ipykernel_19600\559104121.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_1, connection)


Количество вакансий
0                  351

5. Сколько ключевых навыков в среднем указывают в вакансиях для DS?
Ответ округлите до двух знаков после точки-разделителя.

In [14]:
query = f'''SELECT
ROUND((AVG(length(key_skills) - length(replace(key_skills, CHR(9), ''))+1)),2)


FROM public.VACANCIES 

WHERE (lower(name) LIKE '%data scientist%' 
OR lower(name) LIKE '%data science%'
OR lower(name) LIKE '%исследователь данных%'
OR (name LIKE '%ML%' AND name NOT LIKE '%HTML%')
OR lower(name) LIKE '%machine learning%'
OR lower(name) LIKE '%машинн%обучен%')
AND (key_skills IS NOT null)



'''

In [15]:
df = pd.read_sql_query(query, connection)
df 

C:\Users\COMP\AppData\Local\Temp\ipykernel_19600\559104121.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_1, connection)


round
0   6.41

6. Напишите запрос, позволяющий вычислить, какую зарплату для DS в **среднем** указывают для каждого типа требуемого опыта (уникальное значение из поля *experience*). 

При решении задачи примите во внимание следующее:
1. Рассматриваем только вакансии, у которых заполнено хотя бы одно из двух полей с зарплатой.
2. Если заполнены оба поля с зарплатой, то считаем зарплату по каждой вакансии как сумму двух полей, делённую на 2. Если заполнено только одно из полей, то его и считаем зарплатой по вакансии.
3. Если в расчётах участвует null, в результате он тоже даст null (посмотрите, что возвращает запрос select 1 + null). Чтобы избежать этой ситуацию, мы воспользуемся функцией [coalesce](https://postgrespro.ru/docs/postgresql/9.5/functions-conditional#functions-coalesce-nvl-ifnull), которая заменит null на значение, которое мы передадим. Например, посмотрите, что возвращает запрос `select 1 + coalesce(null, 0)`

Выясните, на какую зарплату в среднем может рассчитывать дата-сайентист с опытом работы от 3 до 6 лет. Результат округлите до целого числа. 

In [20]:
query = f'''with result AS (SELECT 
experience AS experience,
coalesce((salary_from+salary_to)/2, salary_from, salary_to) AS salary

FROM public.VACANCIES 

WHERE (lower(name) LIKE '%data scientist%' 
OR lower(name) LIKE '%data science%'
OR lower(name) LIKE '%исследователь данных%'
OR (name like '%ML%' AND name NOT LIKE '%HTML%')
OR lower(name) LIKE '%machine learning%'
OR lower(name) LIKE '%машинн%обучен%')
)

SELECT
experience AS "Опыт работы",
ROUND(AVG(salary)) AS "Средняя зарплата"

FROM result
GROUP BY experience

'''


In [21]:
df = pd.read_sql_query(query, connection)
df

C:\Users\COMP\AppData\Local\Temp\ipykernel_19600\3851344990.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_1, connection)


Опыт работы  Средняя зарплата
0       От 3 до 6 лет          243115.0
1  От 1 года до 3 лет          139675.0
2           Нет опыта           74643.0
3         Более 6 лет               NaN

***

Вакансии в количестве 1771 размещены на сайте имеющие отношения к данным. Вакансий для начинающего дата-сайентиста размещено на сайте в количестве 51. Вакансий с ключевым навыками SQL или postgres размещено на сайте в количестве 201.  Вакансий с требованием владения Python размещено в количестве 351. В среднем вакансии по поиску дата-сайентиста имеют ключевые навыки в количестве 6.41.Чем больше у человека опыта работы, тем больше он получает зарплату.

# Дополнительное исследование №1

 Хочу узнать почему Яндекс разместил вакансии в большем количестве регионов, чем другие работодатели.

In [23]:
query = f'''SELECT
e.name,
schedule||' — '||employment AS "Сочетание",
COUNT(schedule||' — '||employment) AS "Количество"

FROM public.VACANCIES AS v
    JOIN public.EMPLOYERS AS e
    ON e.id = v.employer_id

WHERE e.name LIKE '%Яндекс'
GROUP BY e.name,employment, schedule
ORDER BY "Количество" desc
'''

In [24]:
df = pd.read_sql_query(query, connection)
df

C:\Users\COMP\AppData\Local\Temp\ipykernel_24600\3851344990.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_1, connection)


name                               Сочетание  Количество
0  Яндекс  Удаленная работа — Частичная занятость        1045
1  Яндекс     Удаленная работа — Полная занятость         645
2  Яндекс          Полный день — Полная занятость         125
3  Яндекс    Сменный график — Частичная занятость          55
4  Яндекс     Гибкий график — Частичная занятость          35
5  Яндекс        Гибкий график — Полная занятость          14
6  Яндекс                Полный день — Стажировка           6
7  Яндекс              Гибкий график — Стажировка           6
8  Яндекс       Сменный график — Полная занятость           2

В связи с тем, что у Яндекса в основном вакансии предусматривают удалённый тип рыботы. Соответственно им необязатеьно в каждом регионе открывать офисы. Чтобы больше сотрудников привлечь необходимо создавать условия с удалённым типом работы.

# Дополнительное исследование №2

Какой из работодателей указал зарплату выше остальных

In [137]:
query = f'''with result AS (SELECT 
e.name AS "Работодатель",
coalesce((salary_from+salary_to)/2, salary_from, salary_to) AS salary,
v.name AS "Название вакансии",
v.schedule AS "Тип рабочего графика",
employment AS "Тип трудоустройства",
a.name AS "Регион",
v.experience AS "Требование к опыту"


FROM public.VACANCIES AS v
    JOIN public.EMPLOYERS AS e
    ON e.id = v.employer_id
    
    JOIN public.AREAS AS a
    ON a.id = v.area_id

WHERE (lower(v.name) LIKE '%data scientist%' 
OR lower(v.name) LIKE '%data science%'
OR lower(v.name) LIKE '%исследователь данных%'
OR (v.name like '%ML%' AND v.name NOT LIKE '%HTML%')
OR lower(v.name) LIKE '%machine learning%'
OR lower(v.name) LIKE '%машинн%обучен%')

)

SELECT
"Работодатель",
ROUND(AVG(salary)) AS "Средняя зарплата",
"Название вакансии",
"Тип рабочего графика",
"Тип трудоустройства",
"Требование к опыту",
"Регион" 

FROM result

WHERE salary IS NOT NULL 

GROUP BY "Работодатель", "Название вакансии", "Тип рабочего графика", "Тип трудоустройства",
"Регион", "Требование к опыту"

ORDER BY "Средняя зарплата" DESC
LIMIT 1
'''

In [164]:
df = pd.read_sql_query(query, connection)
df


C:\Users\COMP\AppData\Local\Temp\ipykernel_24600\4206390918.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_1, connection)


Работодатель  Средняя зарплата             Название вакансии  \
0         Итон          450000.0  ML Engineer \ Data scientist   

  Тип рабочего графика Тип трудоустройства Требование к опыту  Регион  
0          Полный день    Полная занятость      От 3 до 6 лет  Москва

Работодатель "Итон" указал в вакансии самую высокую среднюю зарплату. Название вакансии 'ML Engineer \ Data scientist'.Полный рабочий день в офисе в Москве.Стаж у сотрудника должен быть не менее трёх лет.

# Общий вывод по проекту

По итогам исследования можно сделать вывод:
1.Зарплата выше всего в Москве;
2.Вакансий больше всего в Москве;
3.Чем больше стаж, тем выше зарплата;
4.Более 80% вакансий с типом трудоустройства "полный день — полная занятость";
5.Чем выше требуемый стаж, тем меньше вакансий. Это связано скоре всего с тем, что работодатель отбирает более опытных специалистов внутри компании или опытные специалисты сами отправляют резюме потенциальному работодателю.

Было бы большим преимуществом для работодателя, если бы он практиковал тип рабочего графика "удалённая работа". Так работадатель бы смог охватить работников со всего мира.Отбирал бы лучших из лучших, ведь не у каждого человека есть возможность переехать и жить к примеру в Москве.Жить из-за работы в городе не каждому будет в удовольствие.